packages/compiler/server/serverlib.ts

```diff
@@ -403,6 +408,44 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function getFoldingRanges(params: FoldingRangeParams): Promise<FoldingRange[]> {
+    const file = await compilerHost.readFile(getPath(params.textDocument));
+    const ast = parse(file);
+    const ranges: FoldingRange[] = [];
+    getRangesForNode(ast);
+
+    function getRangesForNode(node: Node) {
+      let startPos = node.pos;
+      let endPos = node.end;
+      if ("decorators" in node && node.decorators.length > 0) {
+        startPos = node.decorators[0].pos;
+        endPos = node.decorators[node.decorators.length - 1].end;
+      }
+      let startLine = file.getLineAndCharacterOfPosition(startPos).line;
+      let endLine = file.getLineAndCharacterOfPosition(endPos).line;
+      const startCharacter = file.getLineAndCharacterOfPosition(startPos).character;
+      const endCharacter = file.getLineAndCharacterOfPosition(endPos).character;
+      if ("decorators" in node && node.decorators.length > 0) {
+        const nodeEndPos = node.end;
+        if (startLine !== endLine) {
+          ranges.push({ startLine, startCharacter, endLine, endCharacter });
+        }
+        startLine = file.getLineAndCharacterOfPosition(skipTrivia(file.text, endPos)).line;
```

Comment:
The skipTrivia reminds me that we should file an issue to fold comments. Can you open that?

We want to be able to collapse consecutive lines of // and /** to **/ ranges. Comments aren't nodes so you will have to handle them separately by passing includeComments option to parse and using the comments array on the root of the AST that are just positions, not nodes that can be visited.

All of this for info on that next issue, not as feedback for this PR. 😀

Reply:
sure https://github.com/microsoft/cadl/issues/690